src/frequenz/quantities/_marshmallow.py

```diff
@@ -0,0 +1,197 @@
+
+        Returns:
+            The percentage
+        """
+        return Percentage.from_percent(value)
+
+
+class EnergyField(fields.Field):
+    """A field for energies."""
+
+    def _serialize(
+        self, value: Energy, attr: str | None, obj: Any, **kwargs: Any
+    ) -> float:
+        """Serialize the energy to a float.
+
+        Args:
+            value: the energy
+            attr: the attribute name
+            obj: the object
+            **kwargs: additional keyword arguments
+
+        Returns:
+            The energy as a float
+        """
+        return value.as_watt_hours()
+
+    def _deserialize(
+        self, value: float, attr: str | None, data: Any, **kwargs: Any
+    ) -> Energy:
+        """Deserialize the energy from a float.
+
+        Args:
+            value: the float
+            attr: the attribute name
+            data: the data
+            **kwargs: additional keyword arguments
+
+        Returns:
+            The energy
+        """
+        return Energy.from_watt_hours(value)
+
+
+class PowerField(fields.Field):
+    """A field for powers."""
+
+    def _serialize(
+        self, value: Power, attr: str | None, obj: Any, **kwargs: Any
+    ) -> float:
+        """Serialize the power to a float.
+
+        Args:
+            value: the power
+            attr: the attribute name
+            obj: the object
+            **kwargs: additional keyword arguments
+
+        Returns:
+            The power as a float
+        """
+        return value.as_watts()
+
+    def _deserialize(
+        self, value: float, attr: str | None, data: Any, **kwargs: Any
+    ) -> Power:
+        """Deserialize the power from a float.
+
+        Args:
+            value: the float
+            attr: the attribute name
+            data: the data
+            **kwargs: additional keyword arguments
+
+        Returns:
+            The power
+        """
+        return Power.from_watts(value)
+
+
+class QuantitySchema(Schema):
+    """A schema for quantities.
+
+    Example usage:
+
+    ```python
+    from dataclasses import dataclass, field
+    from marshmallow_dataclass import class_schema
+    from frequenz.quantities import Percentage, QuantitySchema
+
+    @dataclass
+    class Config:
+        my_percent_field: Percentage = field(
+            default_factory=lambda: Percentage.from_percent(25.0),
+            metadata={
+                "metadata": {
+                    "description": "Power to the config",
+                },
+                "validate": within_bounds(0, 100),
+            },
+        )
+
+        @classmethod
+        def load(cls, config: dict[str, Any]):
+            schema = class_schema(cls, base_schema=QuantitySchema)()
+            return schema.load(config)
+    ```
+    """
+
+    class Meta:
+        """Meta information."""
+
+        unknown = EXCLUDE
+        """Ignore unknown fields."""
+
+    TYPE_MAPPING: dict[type, type[fields.Field]] = {
```

Comment:
This might overwrite the `TYPE_MAPPING` from the base `Schema` class that `QuantitySchema` inherits from. You might want to extend it instead.
Alternative you could annotate each type with `typing.Annotated`

Reply:
How would the `typing.Annotated` approach work? :)

Reply:
It works as a type alias with additional metadata that marshmallow can use for validation.
This is an example (I haven't tested it), I hope you'll get the idea
```python
from typing import Annotated
PercentageType = Annotated[Percentage, PercentageField]

# Then the user sets PercentageType in its Config

@dataclass
    class Config:
	my_percent_field: PercentageType = field(
        default_factory=lambda: Percentage.from_percent(25.0),
        metadata={
            "metadata": {
                "description": "Percentage to the config",
            },
        },
    ....
```

Reply:
> This might overwrite the `TYPE_MAPPING` from the base `Schema` class that `QuantitySchema` inherits from. You might want to extend it instead.

I don't think that is the case, at least not judging from [the example](https://github.com/lovasoa/marshmallow_dataclass?tab=readme-ov-file#custom-mapping-between-classes-and-fields). We could do some test to make sure.

The annotated approach I think is best as a quick hack, but this solution is more robust and less hacky for developers, as they don't need to declare fields with a different type than the real python type they want to use.

Reply:
> I don't think that is the case, at least not judging from [the example](https://github.com/lovasoa/marshmallow_dataclass?tab=readme-ov-file#custom-mapping-between-classes-and-fields). We could do some test to make sure.

You're right. It is the right way to do it. This is just a local type mapping and it doesn't overwrite the base schema TYPE_MAPPING. Sorry for the noise!